File: go/mechanisms/evm/batch-settlement/server/file_storage.go
package server

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	batchsettlement "github.com/x402-foundation/x402/go/v2/mechanisms/evm/batch-settlement"
)

// FileChannelStorage is a file-backed SessionStorage. Each session is stored
// as {root}/server/{channelId}.json. CompareAndSet is serialised through an
// exclusive lock file ({channelId}.json.lock) so concurrent writers see the
// loser as a no-op rather than racing.
type FileChannelStorage struct {
	root string
}

// NewFileChannelStorage returns a file-backed server session storage.
func NewFileChannelStorage(opts batchsettlement.FileChannelStorageOptions) *FileChannelStorage {
	return &FileChannelStorage{root: opts.Directory}
}

func (s *FileChannelStorage) filePath(channelId string) (string, error) {
	id, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return "", err
	}
	return batchsettlement.ResolveWithinDir(filepath.Join(s.root, "server"), id+".json")
}

func (s *FileChannelStorage) holdPath(channelId string) (string, error) {
	id, err := batchsettlement.NormalizeChannelId(channelId)
	if err != nil {
		return "", err
	}
	return batchsettlement.ResolveWithinDir(filepath.Join(s.root, "server"), id+".hold")
}

func (s *FileChannelStorage) Get(channelId string) (*ChannelSession, error) {
	path, err := s.filePath(channelId)
	if err != nil {
		return nil, err
	}
	out := &ChannelSession{}
	ok, err := batchsettlement.ReadJSONFile(path, out)
	if err != nil {
		return nil, err
	}
	if !ok {
		return nil, nil
	}
	return out, nil
}

func (s *FileChannelStorage) Set(channelId string, session *ChannelSession) error {
	path, err := s.filePath(channelId)
	if err != nil {
		return err
	}
	return batchsettlement.WriteJSONAtomic(path, session)
}

func (s *FileChannelStorage) Delete(channelId string) error {
	path, err := s.filePath(channelId)
	if err != nil {
		return err
	}
	if err := os.Remove(path); err != nil && !batchsettlement.IsNotExist(err) {
		return err
	}
	return s.dropHold(channelId)
}

func (s *FileChannelStorage) List() ([]*ChannelSession, error) {
	dir := filepath.Join(s.root, "server")
	entries, err := os.ReadDir(dir)
	if err != nil {
		if batchsettlement.IsNotExist(err) {
			return []*ChannelSession{}, nil
		}
		return nil, err
	}

	sessions := make([]*ChannelSession, 0, len(entries))
	for _, entry := range entries {
		name := entry.Name()
		if !strings.HasSuffix(name, ".json") || strings.HasSuffix(name, ".lock") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(dir, name))
		if err != nil {
			// Skip files that disappeared between readdir and read (concurrent delete).
			if batchsettlement.IsNotExist(err) {
				continue
			}
			return nil, err
		}
		out := &ChannelSession{}
		if err := json.Unmarshal(raw, out); err != nil {
			return nil, fmt.Errorf("unmarshal %s: %w", name, err)
		}
		sessions = append(sessions, out)
	}
	sort.Slice(sessions, func(i, j int) bool { return sessions[i].ChannelId < sessions[j].ChannelId })
	return sessions, nil
}

// CompareAndSet uses an exclusive lock file to serialise concurrent writers.
// The mkdir call ensures the very first CompareAndSet on a fresh directory
// does not fail with ENOENT on the lock file.
func (s *FileChannelStorage) CompareAndSet(channelId string, expectedCharged string, session *ChannelSession) (bool, error) {
	path, err := s.filePath(channelId)
	if err != nil {
		return false, err
	}
	lockPath := path + ".lock"

	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return false, fmt.Errorf("mkdir %s: %w", filepath.Dir(lockPath), err)
	}

	lockFile, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
	if err != nil {
		if errors.Is(err, os.ErrExist) {
			return false, nil
		}
		return false, fmt.Errorf("acquire lock %s: %w", lockPath, err)
	}
	defer func() {
		_ = lockFile.Close()
		_ = os.Remove(lockPath)
	}()

	current := &ChannelSession{}
	ok, err := batchsettlement.ReadJSONFile(path, current)
	if err != nil {
		return false, err
	}
	if ok && current.ChargedCumulativeAmount != expectedCharged {
		return false, nil
	}
	if err := batchsettlement.WriteJSONAtomic(path, session); err != nil {
		return false, err
	}
	return true, nil
}

// UpdateChannel atomically reads, mutates, and writes a channel record under an
// exclusive lock file. Returning a different pointer commits the new session;
// returning nil deletes the file; returning the same pointer is treated as a
// no-op (status: unchanged).
func (s *FileChannelStorage) UpdateChannel(channelId string, update func(current *ChannelSession) *ChannelSession) (*ChannelUpdateResult, error) {
	path, err := s.filePath(channelId)
	if err != nil {
		return nil, err
	}
	lockPath := path + ".lock"

	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return nil, fmt.Errorf("mkdir %s: %w", filepath.Dir(lockPath), err)
	}

	lockFile, err := acquireExclusiveFile(lockPath, nil)
	if err != nil {
		return nil, err
	}
	defer func() {
		_ = lockFile.Close()
		_ = os.Remove(lockPath)
	}()

	var current *ChannelSession
	out := &ChannelSession{}
	ok, err := batchsettlement.ReadJSONFile(path, out)
	if err != nil {
		return nil, err
	}
	if ok {
		current = out
	}

	next := update(current)
	switch {
	case next == nil:
		if !ok {
			return &ChannelUpdateResult{Status: ChannelUnchanged}, nil
		}
		if rmErr := os.Remove(path); rmErr != nil && !batchsettlement.IsNotExist(rmErr) {
			return nil, rmErr
		}
		if dropErr := s.dropHold(channelId); dropErr != nil {
			return nil, dropErr
		}
		return &ChannelUpdateResult{Status: ChannelDeleted}, nil
	case current != nil && next == current:
		return &ChannelUpdateResult{Channel: current, Status: ChannelUnchanged}, nil
	default:
		if err := batchsettlement.WriteJSONAtomic(path, next); err != nil {
			return nil, err
		}
		return &ChannelUpdateResult{Channel: next, Status: ChannelUpdated}, nil
	}
}

func (s *FileChannelStorage) Acquire(channelId string, pendingId string, ttlMs int64) (bool, error) {
	var acquired bool
	err := s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
			return fmt.Errorf("mkdir %s: %w", filepath.Dir(path), err)
		}
		raw, readErr := os.ReadFile(path)
		if readErr != nil && !batchsettlement.IsNotExist(readErr) {
			return readErr
		}
		if readErr == nil {
			var existing admissionLock
			if unmarshalErr := json.Unmarshal(raw, &existing); unmarshalErr != nil {
				return unmarshalErr
			}
			if existing.ExpiresAt > time.Now().UnixMilli() {
				acquired = false
				return nil
			}
		}
		record, err := json.Marshal(admissionLock{PendingId: pendingId, ExpiresAt: time.Now().UnixMilli() + ttlMs})
		if err != nil {
			return err
		}
		if err := os.WriteFile(path, record, 0o644); err != nil {
			return err
		}
		acquired = true
		return nil
	})
	return acquired, err
}

func (s *FileChannelStorage) Release(channelId string, pendingId string) error {
	return s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		raw, err := os.ReadFile(path)
		if err != nil {
			if batchsettlement.IsNotExist(err) {
				return nil
			}
			return err
		}
		var hold admissionLock
		if err := json.Unmarshal(raw, &hold); err != nil {
			return err
		}
		if hold.PendingId != pendingId {
			return nil
		}
		if rmErr := os.Remove(path); rmErr != nil && !batchsettlement.IsNotExist(rmErr) {
			return rmErr
		}
		return nil
	})
}

func (s *FileChannelStorage) IsHeld(channelId string, pendingId string) (bool, error) {
	var held bool
	err := s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		raw, err := os.ReadFile(path)
		if err != nil {
			if batchsettlement.IsNotExist(err) {
				held = false
				return nil
			}
			return err
		}
		var hold admissionLock
		if err := json.Unmarshal(raw, &hold); err != nil {
			return err
		}
		if hold.ExpiresAt <= time.Now().UnixMilli() {
			held = false
			return nil
		}
		if pendingId == "" {
			held = true
			return nil
		}
		held = hold.PendingId == pendingId
		return nil
	})
	return held, err
}

func (s *FileChannelStorage) dropHold(channelId string) error {
	return s.withHoldLock(channelId, func() error {
		path, err := s.holdPath(channelId)
		if err != nil {
			return err
		}
		if rmErr := os.Remove(path); rmErr != nil && !batchsettlement.IsNotExist(rmErr) {
			return rmErr
		}
		return nil
	})
}

// withHoldLock serializes Acquire, Release, and IsHeld on {id}.hold.lock so an
// expired hold cannot be unlinked out from under a new holder.
func (s *FileChannelStorage) withHoldLock(channelId string, fn func() error) error {
	path, err := s.holdPath(channelId)
	if err != nil {
		return err
	}
	lockPath := path + ".lock"
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		return fmt.Errorf("mkdir %s: %w", filepath.Dir(lockPath), err)
	}
	lockFile, err := acquireExclusiveFile(lockPath, nil)
	if err != nil {
		return err
	}
	defer func() {
		_ = lockFile.Close()
		_ = os.Remove(lockPath)
	}()
	return fn()
}

type acquireExclusiveFileOptions struct {
	maxAttempts   int
	retryInterval time.Duration
	staleAfter    time.Duration
}

func (o *acquireExclusiveFileOptions) withDefaults() acquireExclusiveFileOptions {
	out := *o
	if out.maxAttempts <= 0 {
		out.maxAttempts = 50
	}
	if out.retryInterval <= 0 {
		out.retryInterval = 10 * time.Millisecond
	}
	if out.staleAfter <= 0 {
		out.staleAfter = 2 * time.Second
	}
	return out
}

// acquireExclusiveFile creates lockPath with O_EXCL, polling until the marker is
// free or stale (mtime older than staleAfter). Stale markers are unlinked so a
// crash cannot pin the channel forever.
func acquireExclusiveFile(lockPath string, opts *acquireExclusiveFileOptions) (*os.File, error) {
	var cfg acquireExclusiveFileOptions
	if opts != nil {
		cfg = opts.withDefaults()
	} else {
		cfg = (&acquireExclusiveFileOptions{}).withDefaults()
	}
	for attempt := 0; attempt < cfg.maxAttempts; attempt++ {
		f, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
		if err == nil {
			return f, nil
		}
		if !errors.Is(err, os.ErrExist) {
			return nil, fmt.Errorf("acquire lock %s: %w", lockPath, err)
		}
		info, statErr := os.Stat(lockPath)
		if statErr != nil {
			if batchsettlement.IsNotExist(statErr) {
				continue
			}
			return nil, fmt.Errorf("acquire lock %s: %w", lockPath, statErr)
		}
		if time.Since(info.ModTime()) > cfg.staleAfter {
			_ = os.Remove(lockPath)
			continue
		}
		time.Sleep(cfg.retryInterval)
	}
	return nil, fmt.Errorf("acquire lock %s: contended", lockPath)
}

File: typescript/packages/mechanisms/evm/src/batch-settlement/server/fileStorage.ts
import { mkdir, open, readdir, readFile, stat, unlink, writeFile } from "node:fs/promises";
import { constants } from "node:fs";
import { dirname, join } from "node:path";

import { isNodeEnoent, readJsonFile, resolveWithinDir, writeJsonAtomic } from "../storage-utils";
import { normalizeChannelId } from "../utils";
import type { FileChannelStorageOptions } from "../types";
import type { ChannelLockStorage, ChannelStorage, Channel, ChannelUpdateResult } from "./storage";

export type { FileChannelStorageOptions };

const FILE_LOCK_MAX_ATTEMPTS = 50;
const FILE_LOCK_RETRY_INTERVAL_MS = 10;
const FILE_LOCK_STALE_MS = 2_000;

export type AcquireExclusiveFileOptions = {
  maxAttempts?: number;
  retryIntervalMs?: number;
  staleMs?: number;
};

/**
 * Node.js file-backed {@link ChannelStorage} for the batched server scheme.
 */
export class FileChannelStorage implements ChannelStorage, ChannelLockStorage {
  private readonly root: string;

  /**
   * Creates file-backed server channel storage under the given root directory.
   *
   * @param options - Configuration including the storage root directory.
   */
  constructor(options: FileChannelStorageOptions) {
    this.root = options.directory;
  }

  /**
   * Loads a persisted channel record, if present.
   *
   * @param channelId - The channel identifier (path segment is lowercased).
   * @returns Parsed channel record or `undefined` when the file is missing.
   */
  async get(channelId: string): Promise<Channel | undefined> {
    return readJsonFile<Channel>(this.filePath(channelId));
  }

  /**
   * Lists all stored channel records by reading the server directory.
   *
   * @returns Channel records sorted by channelId; empty array if the directory is missing.
   */
  async list(): Promise<Channel[]> {
    const dir = join(this.root, "server");
    let names: string[];
    try {
      names = await readdir(dir);
    } catch (err: unknown) {
      if (isNodeEnoent(err)) return [];
      throw err;
    }

    const channels: Channel[] = [];
    for (const name of names) {
      if (!name.endsWith(".json")) continue;
      const path = join(dir, name);
      try {
        const raw = await readFile(path, "utf8");
        channels.push(JSON.parse(raw) as Channel);
      } catch (err: unknown) {
        // Skip files that disappeared between readdir and readFile (e.g. concurrent delete).
        // Rethrow other failures (corrupt JSON, permission denied) so callers see them.
        if (isNodeEnoent(err)) continue;
        throw err;
      }
    }
    return channels.sort((a, b) => a.channelId.localeCompare(b.channelId));
  }

  /**
   * Atomically inspects and mutates a channel record under a cross-process file lock.
   *
   * @param channelId - The channel identifier.
   * @param update - Mutation callback. Return `undefined` to delete, or `current` to leave unchanged.
   * @returns The final stored channel and whether storage updated, stayed unchanged, or deleted.
   */
  async updateChannel(
    channelId: string,
    update: (current: Channel | undefined) => Channel | undefined,
  ): Promise<ChannelUpdateResult> {
    const lockPath = this.filePath(channelId) + ".lock";
    await mkdir(dirname(lockPath), { recursive: true });
    const lockHandle = await acquireExclusiveFile(lockPath);

    try {
      const path = this.filePath(channelId);
      let current: Channel | undefined;
      try {
        const raw = await readFile(path, "utf8");
        current = JSON.parse(raw) as Channel;
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }

      const next = update(current);
      if (next === current) {
        return { channel: current, status: "unchanged" };
      }

      if (!next) {
        try {
          await unlink(path);
        } catch (err: unknown) {
          if (!isNodeEnoent(err)) throw err;
        }
        await this.dropHold(channelId);
        return { channel: undefined, status: current ? "deleted" : "unchanged" };
      }

      await writeJsonAtomic(path, next);
      return { channel: next, status: "updated" };
    } finally {
      await lockHandle.close();
      await unlink(lockPath).catch(() => {});
    }
  }

  /**
   * Acquires a per-channel admission lock via a sidecar hold file.
   *
   * Serialized with {@link FileChannelStorage.release} and {@link FileChannelStorage.isHeld}
   * on `{id}.hold.lock` so an expired hold cannot be unlinked out from under a new holder.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - Request-scoped lock owner.
   * @param ttlMs - Lock time-to-live in milliseconds.
   * @returns Whether this request now holds the lock.
   */
  async acquire(channelId: string, pendingId: string, ttlMs: number): Promise<boolean> {
    return this.withHoldLock(channelId, async () => {
      const path = this.holdPath(channelId);
      await mkdir(dirname(path), { recursive: true });
      try {
        const existing = JSON.parse(await readFile(path, "utf8")) as {
          pendingId: string;
          expiresAt: number;
        };
        if (existing.expiresAt > Date.now()) return false;
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }
      await writeFile(path, JSON.stringify({ pendingId, expiresAt: Date.now() + ttlMs }), "utf8");
      return true;
    });
  }

  /**
   * Releases the admission lock only when `pendingId` still holds it.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - Request-scoped lock owner.
   */
  async release(channelId: string, pendingId: string): Promise<void> {
    await this.withHoldLock(channelId, async () => {
      const path = this.holdPath(channelId);
      try {
        const hold = JSON.parse(await readFile(path, "utf8")) as { pendingId: string };
        if (hold.pendingId !== pendingId) return;
        await unlink(path);
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }
    });
  }

  /**
   * Returns whether a live admission lock exists, optionally matching `pendingId`.
   *
   * @param channelId - The channel identifier.
   * @param pendingId - When set, require this request to hold the lock.
   * @returns Whether a live lock (or this request's lock) is present.
   */
  async isHeld(channelId: string, pendingId?: string): Promise<boolean> {
    return this.withHoldLock(channelId, async () => {
      try {
        const hold = JSON.parse(await readFile(this.holdPath(channelId), "utf8")) as {
          pendingId: string;
          expiresAt: number;
        };
        if (hold.expiresAt <= Date.now()) return false;
        return pendingId === undefined || hold.pendingId === pendingId;
      } catch (err: unknown) {
        if (isNodeEnoent(err)) return false;
        throw err;
      }
    });
  }

  /**
   * Absolute path to the JSON file for a channel.
   *
   * @param channelId - The channel identifier.
   * @returns Filesystem path under `{root}/server/...`.
   * @throws When `channelId` is not a canonical `bytes32` string or escapes the storage root.
   */
  private filePath(channelId: string): string {
    const id = normalizeChannelId(channelId);
    return resolveWithinDir(join(this.root, "server"), `${id}.json`);
  }

  /**
   * Absolute path to the admission hold sidecar for a channel.
   *
   * @param channelId - The channel identifier.
   * @returns Filesystem path under `{root}/server/{id}.hold`.
   */
  private holdPath(channelId: string): string {
    const id = normalizeChannelId(channelId);
    return resolveWithinDir(join(this.root, "server"), `${id}.hold`);
  }

  /**
   * Drops the admission hold for a deleted channel row.
   *
   * @param channelId - The channel identifier.
   */
  private async dropHold(channelId: string): Promise<void> {
    await this.withHoldLock(channelId, async () => {
      try {
        await unlink(this.holdPath(channelId));
      } catch (err: unknown) {
        if (!isNodeEnoent(err)) throw err;
      }
    });
  }

  /**
   * Serializes acquire, release, and isHeld on the `.hold` sidecar.
   *
   * @param channelId - The channel identifier.
   * @param fn - Work to run while holding `{id}.hold.lock`.
   * @returns The resolved result of `fn`.
   */
  private async withHoldLock<T>(channelId: string, fn: () => Promise<T>): Promise<T> {
    const lockPath = this.holdPath(channelId) + ".lock";
    await mkdir(dirname(lockPath), { recursive: true });
    const lockHandle = await acquireExclusiveFile(lockPath);
    try {
      return await fn();
    } finally {
      await lockHandle.close();
      await unlink(lockPath).catch(() => {});
    }
  }
}

/**
 * Creates an exclusive lock file, stealing markers whose mtime is older than `staleMs`.
 *
 * @param lockPath - Absolute path for the lock file (created with `O_EXCL`).
 * @param options - Attempt, retry, and stale-mtime bounds.
 * @returns Writable file handle for the lock file; caller must close it to release.
 * @throws When the lock remains contended after `maxAttempts`.
 */
export async function acquireExclusiveFile(
  lockPath: string,
  options: AcquireExclusiveFileOptions = {},
) {
  const maxAttempts = options.maxAttempts ?? FILE_LOCK_MAX_ATTEMPTS;
  const retryIntervalMs = options.retryIntervalMs ?? FILE_LOCK_RETRY_INTERVAL_MS;
  const staleMs = options.staleMs ?? FILE_LOCK_STALE_MS;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    try {
      return await open(lockPath, constants.O_CREAT | constants.O_EXCL | constants.O_WRONLY);
    } catch (err: unknown) {
      if ((err as NodeJS.ErrnoException).code !== "EEXIST") {
        throw err;
      }
      if (await isStaleLock(lockPath, staleMs)) {
        await unlink(lockPath).catch(() => {});
        continue;
      }
      await new Promise(resolve => setTimeout(resolve, retryIntervalMs));
    }
  }
  throw new Error(`acquire lock ${lockPath}: contended`);
}

/**
 * Returns whether `lockPath` is missing or older than `staleMs`.
 *
 * @param lockPath - Absolute path for the lock file.
 * @param staleMs - Age after which a leftover marker is treated as crash debris.
 * @returns Whether the caller may unlink and retry exclusive create.
 */
async function isStaleLock(lockPath: string, staleMs: number): Promise<boolean> {
  try {
    const info = await stat(lockPath);
    return Date.now() - info.mtimeMs >= staleMs;
  } catch (err: unknown) {
    if (isNodeEnoent(err)) return true;
    throw err;
  }
}
